Update rust-canister-development-security-best-practices.md (#1100)
dfinity/portal: docs/developer-docs/security/rust-canister-development-security-best-practices.md

- See [Current limitations of the Internet Computer](https://wiki.internetcomputer.org/wiki/Current_limitations_of_the_Internet_Computer), section "Bugs in `pre_upgrade` hooks"

> ### Reinstantiate global timers during upgrades

Reviewer: Would it make sense to drop the word "global" because it sounds confusing?

On the system API level, we have a single global timer (that's why it is called global). Motoko and Rust CDK implement support for many user timers on top of the global timer.

Author: ```suggestion
### Reinstantiate timers during upgrades
```

#### Security Concern

> Global timers are deactivated upon changes to the canister's Wasm module. The [IC specification](https://internetcomputer.org/docs/current/references/ic-interface-spec#timer) states this as follows:

Reviewer 2: ```suggestion
Global timers are deactivated upon changes to the canister's Wasm module. The [IC specification](../references/ic-interface-spec#timer) states this as follows:
```
relative links, please! otherwise they don't get checked by CI

> > The global timer is also deactivated upon changes to the canister's Wasm module (calling install_code, uninstall_code methods of the management canister or if the canister runs out of cycles). In particular, the function canister_global_timer won't be scheduled again unless the canister sets the global timer again (using the System API function ic0.global_timer_set).

Author: ```suggestion
> The timer is also deactivated upon changes to the canister's Wasm module (calling install_code, uninstall_code methods of the management canister or if the canister runs out of cycles). In particular, the function canister_global_timer won't be scheduled again unless the canister sets the global timer again (using the System API function ic0.global_timer_set).
```

> Upgrade is a mode of `install_code` and hence the global timers are deactivated during an upgrade.

Reviewer 3: Since this describes a "Security Concern", would it make sense to elaborate more what bad consequences for security this could have? E.g. along these lines: 
```
If such a timer is used to manage regular operations such as cleanups or payments, this can be risky from a security perspective. 
```
Maybe a good example would be what could have gone wrong in ckBTC (I don't remember)?

Author: For example, a timer used to update the exchange rate of a trading pair on a DeX.

Author: ```suggestion
Upgrade is a mode of `install_code` and hence the timers are deactivated during an upgrade.

This could result in a vulnerability in certain cases where security controls or other critical features rely on these timers. For example, a DEX which relies on timers to update the exchange rates of currencies could be vulnerable to arbitraging opportunities if the rates are no longer updated.
```

Since global timers are used internally by the Motoko `Timer` mechanism, the same holds true for Motoko Timer. As explained in the [pull request](https://github.com/dfinity/motoko/pull/3542) under "The upgrade story", the global timer gets jettisoned on upgrade, and the timers need to be set up in the post-upgrade hook.

> As explained in [this pull request](https://github.com/dfinity/motoko/pull/3542) under "Opting out", the behavior is different when using Motoko and implementing `system func timer`. The `timer` function will be called after an upgrade. In case your canister was using timers for recurring tasks, the `timer` function would likely set the global timer again for a later time. However, the time between invocations of `timer` would not be consistent as the upgrade triggered an "unexpected" call to `timer`.

Author: According to the original pull request this would happen always after an upgrade:

> **The initial expiration**
> After the canister starts running, system func timer gets invoked automatically. This also happens after an upgrade.

@ggreif, can you confirm?

> Using the rust CDK, the reccuring timer is also lost on upgrade as explained in the API documentation of [set_timer_interval](https://docs.rs/ic-cdk/0.6.9/ic_cdk/timer/fn.set_timer_interval.html).

Reviewer 2: Any chance we can link to something else than a hard-coded version number?

Author: I checked the latest releases and for some reason it didn't have the timer API. I could change it to 0.6.10 since this seems to be the latest patch with the timer API. I'll put a reminder in my calendar in 3 months to check again if latest has the timer API.

#### Recommendation

- Keep track of global timers in the `pre_upgrade` hook. Store any state in stable variables.

> - Set global timers in the `post_upgrade` hook.

Reviewer: Ditto about the word "global"

Author: ```suggestion
- Set timers in the `post_upgrade` hook.
```

> - See the Motoko documentation on [recurringTimer](https://internetcomputer.org/docs/current/references/motoko-ref/Timer#function-recurringtimer)

Reviewer 2: ```suggestion
- See the Motoko documentation on [recurringTimer](../references/motoko-ref/Timer#function-recurringtimer)
```
